torch/csrc/jit/passes/alias_analysis.cpp

```diff
@@ -930,6 +931,36 @@ bool AliasDb::couldMoveBeforeTopologically(Node* n, Node* movePoint) {
   return tryMove(n, movePoint, MoveSide::BEFORE, /*dryRun=*/true);
 }
 
+bool AliasDb::hasWriters(const at::ArrayRef<Value*>& values) const {
+  return std::any_of(values.begin(), values.end(), [&](Value* value) {
+    return hasWriters(value);
+  });
+}
+
+// Correctness conditions:
+// no values in either set can have writers, and values in both sets
+// cannot escape the current graph scope. Values can escape the current scope
+// by aliasing a graph output or input, or aliasing an input to
+// prim::FunctionCall or MethodCall. FunctionCall and MethodCall are marked as
+// mutating ops, so we do not need to check for them because they are covered
```

Comment:
I forgot to mention `prim::SetAttr` but it follows under the same case as `FunctionCall`. 

I should probably add handling of `prim::GetAttr` as well...

Reply:
One thing to be wary of here is that `hasWriters` doesn't express the intent of this query, so while it may be  correct for todays definition, someone might change hasWriters to mean "does this alias something, and do I see a write to it", without realizing it is being used here to ask if a value escapes. It might be more accurate to query whether the value aliases the heap location directly.

Reply:
Updated